auth/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+        for weight in transfer_data.new_weights.iter() {
+            total_weight += weight.deref();
+        }
+
+        require!(
+            transfer_data.new_threshold > 0 && total_weight >= transfer_data.new_threshold,
+            "Invalid threshold"
+        );
+
+        let new_operators_hash = self.get_operators_hash(
+            &transfer_data.new_operators,
+            &transfer_data.new_weights,
+            &transfer_data.new_threshold,
+        );
+
+        require!(
+            self.epoch_for_hash(&new_operators_hash).is_empty(),
+            "Duplicate operators"
+        );
+
+        let epoch = self.current_epoch().update(|epoch| {
+            *epoch += 1;
+            *epoch
+        });
+
+        self.hash_for_epoch(epoch).set(&new_operators_hash);
+        self.epoch_for_hash(&new_operators_hash).set(epoch);
+
+        self.operatorship_transferred_event(transfer_data);
+    }
+
+    fn validate_signatures(
+        &self,
+        message_hash: ManagedByteArray<KECCAK256_RESULT_LEN>,
+        operators: ManagedVec<Operator<Self::Api>>,
+        weights: ManagedVec<BigUint>,
+        threshold: BigUint,
+        signatures: ManagedVec<ManagedByteArray<ED25519_SIGNATURE_BYTE_LEN>>,
+    ) {
+        let mut weight = BigUint::zero();
+
+        for (operator_index, signature) in signatures.iter().enumerate() {
+            let address = operators.get(operator_index);
+
+            self.crypto().verify_ed25519(
+                address.as_managed_buffer(),
+                message_hash.as_managed_buffer(),
+                signature.as_managed_buffer(),
+            );
+
+            // Check that operators do not repeat
+            require!(
+                operators.find(&address) == Some(operator_index),
+                "Malformed signers"
+            );
+
+            let current_weight = weights.get(operator_index);
+            weight += current_weight.deref();
+
+            if weight >= threshold {
+                return;
+            }
+        }
+
+        sc_panic!("Low signatures weight");
+    }
+
+    fn get_operators_hash(
+        &self,
+        operators: &ManagedVec<Operator<Self::Api>>,
+        weights: &ManagedVec<BigUint>,
+        threshold: &BigUint,
+    ) -> ManagedByteArray<KECCAK256_RESULT_LEN> {
+        let mut encoded = ManagedBuffer::new();
+
+        for operator in operators.iter() {
+            encoded.append(operator.as_managed_buffer());
+        }
+
+        for weight in weights.iter() {
+            encoded.append(&weight.to_bytes_be_buffer());
+        }
+
+        encoded.append(&threshold.to_bytes_be_buffer());
+
+        self.crypto().keccak256(encoded)
+    }
+
+    // TODO: Check for a better way of doing this when ManagedMap is available?
+    fn contains_no_duplicate(&self, operators: &ManagedVec<Operator<Self::Api>>) -> bool {
+        if operators.is_empty() {
+            return false;
+        }
+
+        for iindex in 0..(operators.len() - 1) {
```

Comment:
Right now the `operators.len()` instruction is executed `operators.len() * operators.len()` times. 
Get the `operators.len()` out of the `for`s and use that variable.

Reply:
I have made the modification, however now that ManagedMap is available, should I maybe use that?
It think it would reduce CPU time but increase memory usage. Implementation wise I would have to first call `contains` and then call `put` on the ManagedMap in a for loop.

What do you think?